src/main.c

```diff
@@ -1,20 +1,42 @@
 #include <utils/logger.h>
 #include <queue.h>
+#include <stdatomic.h>
+#include <stdbool.h>
 #include <stdio.h>
 #include <unistd.h>
 #include <signal.h>
 #include <producer.h>
 #include <utils/helper.h>
 
-int run   = 1;
-int ready = 0;
+
+/* Schaufel keeps track of consume and produce states.
+ *
+ * consume_state is statically initialised to true (for convenience).
+ * produce_state is initialised on runtime when at least one producer is up.
+ *
+ * The macros must be used to get & set the values of respective states.
+ *
+ * For p = {consume,produce}_process,
+ * $p_init() sets the state to true
+ * $p_done() sets the state to false
+ * $p_stat() queries the value of the state
+ */
+static volatile atomic_bool consume_state = ATOMIC_VAR_INIT(true);
+static volatile atomic_bool produce_state /* = ATOMIC_VAR_INIT(false) */;
+#define consume_process_init() set_state(&consume_state, true)
+#define consume_process_done() set_state(&consume_state, false)
+#define consume_process_stat() atomic_load(&consume_state)
+#define produce_process_init() set_state(&produce_state, true)
+#define produce_process_done() set_state(&produce_state, false)
+#define produce_process_stat() atomic_load(&produce_state)
```

Comment:
not a fan of such preprocessor abuse.

Reply:
I would not call it abuse. I think it would make the code easier to understand (for future me at least, ymmv).

That said I don't think this makes a big difference, so I'll drop that.